### ISSUE TITLE
Joint training phase

### ISSUE THREAD

**Issue by RunshengZhu** [cvlab-stonybrook/DewarpNet#12]
Thanks for the wonderful work~!
    I've done the separate training of two sub-network(depth prediction + texture mapping) with given  data, the result looks good. But there's no code for joint training in origin paper(4.2), is it possible to add loss function to infer.py ?

**Comment by JasonBoy1** [comment]
![image](https://user-images.githubusercontent.com/33180378/79203617-dfd47800-7e6d-11ea-91a0-1c259389fd34.png)

HI, what setting should be adjusted to the released code? I trained the DewarpNet with the default code, and get such a ridiculous result. Could you tell me how you trained it?

**Comment by sagniklp** [comment]
@RunshengZhu Just combine training code of both networks and backpropagate the joint loss.

**Comment by sagniklp** [comment]
@JasonBoy1 I'm not sure what you did. You may try smoothing the output backward mapping.

**Comment by JasonBoy1** [comment]
> @JasonBoy1 I'm not sure what you did. You may try smoothing the output backward mapping.

Yes, I kept all the settings as the released code, and get this result. I have not changed any dataloader or model archs. And I can get a good result using the released 'pkl' file. I am very confused...

**Comment by sagniklp** [comment]
Please elaborate your training procedure. Otherwise I won't be able to help you.

**Comment by JasonBoy1** [comment]
> Please elaborate your training procedure. Otherwise I won't be able to help you.

I prepared the train.txt and val.txt as the ReadMe. Other settings are exactly same as yours.

**Comment by JasonBoy1** [comment]
Dear author, how many epochs do we need to get a good result? Maybe I have not trained for enough time? I am not sure. But I think it takes weeks to train 100 epochs.

**Comment by RunshengZhu** [comment]
@JasonBoy1 As mentioned in title, training code in repo is first stage, you need to combine two part of network and do joint training. Check the paper for details, good luck!

**Comment by RunshengZhu** [comment]
@sagniklp  Thanks for your reply~ I've figure it out. The result looks reasonable, but still not good enough. One more thing, I noticed that in origin paper, augmentation is applied on captured mesh. So which tool did you used on mesh augmentation? Thanks!

**Comment by JasonBoy1** [comment]
> @JasonBoy1 As mentioned in title, training code in repo is first stage, you need to combine two part of network and do joint training. Check the paper for details, good luck!
 Dear Zhu:
Thank you  for your reply. Maybe I should try the joint training method.
By the way, the backwardmap training is very very slow(About 4 epochs/day). Have you encountered this issue?

**Comment by RunshengZhu** [comment]
@JasonBoy1 About 25 epochs/hour on one GTX1080, looks reasonable

**Comment by sagniklp** [comment]
I'm not sure what you've done till now, so can't say why the results aren't perfect. Please follow the guidelines below: 
Make sure your train/test split doesn't have any meshes in common. Follow this: https://github.com/cvlab-stonybrook/doc3D-dataset/issues/8#issuecomment-546566941
Make sure to do the tight crop and augmentation for WC training.
Train the shape network until convergence and use the gradient loss. Increase the weight of the gradient loss after 15-20 epochs.
Train the backward mapping net until convergence with ground-truth WC as input.
Initialize your joint training with the trained models and do the joint training. 

Mesh augmentation is done while rendering the dataset. It is not part of the training process.

**Comment by RunshengZhu** [comment]
@sagniklp Thank you for reply~I've generate my own doc3D dataset with doc3D-renderer, but I have not found process like mentioned in paper(such as flip, crop small sample then resize...). Please tell me what did I missed, thanks!

**Comment by JasonBoy1** [comment]
> @sagniklp Thank you for reply~I've generate my own doc3D dataset with doc3D-renderer, but I have not found process like mentioned in paper(such as flip, crop small sample then resize...). Please tell me what did I missed, thanks!

Could you share how to install the **bpy** lib?

**Comment by JasonBoy1** [comment]
> I'm not sure what you've done till now, so can't say why the results aren't perfect. Please follow the guidelines below:
> Make sure your train/test split doesn't have any meshes in common. Follow this: [cvlab-stonybrook/doc3D-dataset#8 (comment)](https://github.com/cvlab-stonybrook/doc3D-dataset/issues/8#issuecomment-546566941)
> Make sure to do the tight crop and augmentation for WC training.
> Train the shape network until convergence and use the gradient loss. Increase the weight of the gradient loss after 15-20 epochs.
> Train the backward mapping net until convergence with ground-truth WC as input.
> Initialize your joint training with the trained models and do the joint training.
> 
> Mesh augmentation is done while rendering the dataset. It is not part of the training process.

Thanks, I will try.

**Comment by RunshengZhu** [comment]
@JasonBoy1 Install blender, v2.78 works for me

**Comment by JasonBoy1** [comment]
> @JasonBoy1 Install blender, v2.78 works for me
I used pip install blender, but can neither import **'blender'** or **'bpy'.**

**Comment by JasonBoy1** [comment]
@RunshengZhu I used pip install blender, but can neither import **'blender'** or **'bpy'.**

**Comment by RunshengZhu** [comment]
@JasonBoy1' bpy' is python api for blender, you'd better install blender first. Check official document for details. Looks like you need a tutorial(google/medium/csdn), good luck!

**Comment by JasonBoy1** [comment]
@RunshengZhu 
Sorry for bothering, but I have wasted too much time on searching the tutorials, and failed at last. Could you give me a hand like sharing a link? Thank you very much.

**Comment by RunshengZhu** [comment]
@JasonBoy1 https://download.blender.org/release/

**Comment by JasonBoy1** [comment]
Have found a trick: Using grad_loss in the training of backward map Network would make the results more smooth. Without g_loss in bm net training, the surfaces still looks some what 'wavy'.

**Comment by naoto0804** [comment]
@RunshengZhu Did you succeed in joint training? How did you find a `refinement network`?
